components/tidb_query/src/rpn_expr/impl_cast.rs

```diff
@@ -889,6 +898,138 @@ cast_as_duration!(
 );
 cast_as_duration!(Json, cast_json_as_duration, val.unquote()?.as_bytes());
 
+#[rpn_fn(capture = [ctx, extra])]
+fn cast_int_as_time(
+    ctx: &mut EvalContext,
+    extra: &RpnFnCallExtra<'_>,
+    val: &Option<Int>,
+) -> Result<Option<Time>> {
+    if let Some(val) = *val {
+        // Parse `val` as a `u64`
+        Time::parse_from_i64(
+            ctx,
+            val,
+            extra.ret_field_type.as_accessor().tp().try_into()?,
+            extra.ret_field_type.get_decimal() as i8,
+        )
+        .map(Some)
+        .or_else(|_| {
+            Ok(ctx
+                .handle_invalid_time_error(Error::incorrect_datetime_value(val))
+                .map(|_| None)?)
+        })
+    } else {
+        Ok(None)
+    }
+}
+
+// NOTE: in MySQL, casting `Real` to `Time` should cast `Real` to `Int` first,
+// However, TiDB cast `Real` to `String` and then parse it into a `Time`
+#[rpn_fn(capture = [ctx, extra])]
+fn cast_real_as_time(
+    ctx: &mut EvalContext,
+    extra: &RpnFnCallExtra,
+    val: &Option<Real>,
+) -> Result<Option<Time>> {
+    if let Some(val) = val {
+        // Convert `val` to a string first and then parse it as a float string.
+        Time::parse(
+            ctx,
+            &val.to_string(),
+            extra.ret_field_type.as_accessor().tp().try_into()?,
+            extra.ret_field_type.get_decimal() as i8,
+            // Enable round
+            true,
+        )
+        .map(Some)
+        .or_else(|e| Ok(ctx.handle_invalid_time_error(e).map(|_| None)?))
+    } else {
+        Ok(None)
+    }
+}
+
+#[rpn_fn(capture = [ctx, extra])]
+fn cast_string_as_time(
+    ctx: &mut EvalContext,
+    extra: &RpnFnCallExtra,
+    val: &Option<Bytes>,
+) -> Result<Option<Time>> {
+    if let Some(val) = val {
+        // Convert `val` to a string first and then parse it as a float string.
+        Time::parse(
+            ctx,
+            unsafe { std::str::from_utf8_unchecked(val) },
```

Comment:
Is all data from TiDB guaranteed to be utf8?

Reply:
If it is not, time parser will report error.

Reply:
How about accepting byte for this parse function?

Reply:
OK. I will create another pull request to fix this.